From: Robert Ancell <robert.ancell@canonical.com>
Date: Mon, 6 Apr 2020 10:30:27 +1200
Subject: applications: Launch snap-store if it is installed

TODO: snap-store doesn't yet support launching to specific app details page
https://github.com/ubuntu/app-center/issues/1572

TODO: Launch GNOME Software for Flatpaks since snap-store can't handle those
---
 panels/applications/cc-applications-panel.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/panels/applications/cc-applications-panel.c b/panels/applications/cc-applications-panel.c
index 6663ab5..dfc7150 100644
--- a/panels/applications/cc-applications-panel.c
+++ b/panels/applications/cc-applications-panel.c
@@ -165,6 +165,13 @@ gnome_software_is_installed (void)
   return path != NULL;
 }
 
+static gboolean
+snap_store_is_installed (void)
+{
+  g_autofree gchar *path = g_find_program_in_path ("snap-store");
+  return path != NULL;
+}
+
 /* Callbacks */
 
 static gboolean
@@ -184,6 +191,9 @@ open_software_cb (CcApplicationsPanel *self)
 {
   const gchar *argv[] = { "gnome-software", "--details", "appid", NULL };
 
+  if (snap_store_is_installed ())
+    argv[0] = "snap-store";
+
   if (self->current_app_id == NULL)
     argv[1] = NULL;
   else
@@ -1453,7 +1463,8 @@ update_panel (CcApplicationsPanel *self,
   adw_navigation_page_set_title (self->app_settings_page,
                                  g_app_info_get_display_name (info));
   cc_panel_push_subpage (CC_PANEL (self), self->app_settings_page);
-  gtk_widget_set_visible (GTK_WIDGET (self->view_details_button), gnome_software_is_installed ());
+  gtk_widget_set_visible (GTK_WIDGET (self->view_details_button),
+                          gnome_software_is_installed () || snap_store_is_installed());
 
   g_clear_pointer (&self->current_app_id, g_free);
   g_clear_pointer (&self->current_portal_app_id, g_free);
@@ -1878,7 +1889,8 @@ cc_applications_panel_init (CcApplicationsPanel *self)
 
   gtk_widget_init_template (GTK_WIDGET (self));
 
-  gtk_widget_set_visible (GTK_WIDGET (self->install_button), gnome_software_is_installed ());
+  gtk_widget_set_visible (GTK_WIDGET (self->install_button),
+                          gnome_software_is_installed () || snap_store_is_installed ());
 
   g_signal_connect_object (self->app_listbox, "row-activated",
                            G_CALLBACK (row_activated_cb), self, G_CONNECT_SWAPPED);
